pyomo/core/base/external.py

```diff
@@ -54,18 +56,43 @@ def __init__(self, *args, **kwds):
         self._index = None
 
     def __call__(self, *args):
-        idxs = reversed(six.moves.xrange(len(args)))
-        for i in idxs:
+        args_ = []
+        for i in range(len(args)):
             if type(args[i]) is types.GeneratorType:
-                args = args[:i] + tuple(args[i]) + args[i+1:]
-        return EXPR._ExternalFunctionExpression(self, args)
+                args_.extend(val for val in args[i])
+            else:
+                args_.append(args[i])
+        #
+        # Loop and do two thing:
+        #   1. Wrap non-numeric arguments
+        #   2. See if we have a potentially variable argument
+        #
+        pv = False
+        for i in range(len(args_)):
```

Comment:
this should be `for i,arg in enumerate(args_)`

That will avoid a list (in Python 2 -- since you are not using `six`), plus avoid a bunch of `__getitem__` calls below.

Reply:
In fact, the variable `i` is only used to index `args_`, so it's even better to replace this with `for arg in args:`

Reply:
Hmmm... but the exception needs the variable `i`.